src/Core/Aetherling/Analysis/Phase.hs

```diff
@@ -0,0 +1,264 @@
+--
+-- An unfortunate fact is that SequenceArrayRepack needs to have a
+-- tight relationship with underutilization. This is why I've decided
+-- to include a cps field within the SequenceArrayRepack op itself.
+-- The reason for this is that underutilizing a SequenceArrayRepack
+-- by "skipping" unused clock cycles does not produce the behavior
+-- required by the phase patterns specified above.
+--
+-- (Complicated example incoming).
+--
+-- Example: Suppose we convert 4-sequence of 3-arrays to 3 4-arrays.
+--
+-- ‖ clk 0'‖ clk 1'‖ clk 2'‖ clk 3'‖ Clock cycles (wide columns)
+-- ‖       ‖       ‖       ‖       ‖
+-- ‖ 012   ‖ 345   ‖ 678   ‖ 9AB   ‖ Tokens (in)
+-- ‖ |||   ‖ ||└--┐‖ |||   ‖ |||   ‖
+-- ‖ |||   ‖ |└---┐‖ ||└-┐ ‖ ||└┐  ‖
+-- ‖ |||   ‖ └--┐||‖ |└-┐| ‖ |└┐|  ‖
+-- ‖ ||└-------┐|||‖ └-┐|| ‖ └┐||  ‖
+-- ‖ |└-------┐|||└---┐||| ‖  |||  ‖
+-- ‖ └-------┐|||└---┐|||└---┐|||  ‖
+-- ‖       ‖ ||||  ‖ ||||  ‖ ||||  ‖
+-- ‖ xxxx  ‖ 0123  ‖ 4567  ‖ 89AB  ‖ Tokens (out)
+-- ‖       ‖ 1110  ‖ 1100  ‖ 1000  ‖ Token Latency
+--
+-- Now do the same thing, but spread out over 6 clocks. The input is
+-- 4-sequence over 6 clocks. By the rules above they should come in on
+-- cycles 0, 1, 3, and 4. The 3-sequence output should come cycles 1,
+-- 3, and 5. (Phase is 0, 2, 4; +1 latency).
+--
+-- ‖ clk 0 ‖ clk 1 ‖ clk 2 ‖ clk 3 ‖ clk 4 ‖ clk 5 ‖ Clock cycles
+-- ‖       ‖       ‖       ‖       ‖       ‖       ‖
+-- ‖ 012   ‖ 345   ‖ xxx   ‖ 678   ‖ 9AB   ‖ xxx   ‖ Tokens (in)
+-- ‖       ‖       ‖       ‖       ‖       ‖       ‖
+-- ‖ xxxx  ‖ 0123  ‖ xxxx  ‖ 4567  ‖ xxxx  ‖ 89AB  ‖ Tokens (out)
+-- ‖       ‖ 1110  ‖       ‖ 2200  ‖       ‖ 2111  ‖ Token Latency (real)
+-- ‖       ‖       ‖       ‖       ‖       ‖       ‖
+-- ‖ clk 0'‖ clk 1'‖       ‖ clk 2'‖    clk 3'?    ‖
+--
+-- There's actually no way to match up the 4-clocks-per-sequence
+-- example with the 6-cps example by idling for 2 cycles, as underutil
+-- conceptually does. Notice how input 9AB and output 89AB are on the
+-- same cycle (3') in the original but on different cycles (4, 5) in
+-- the underutilized version, and latency can not be reduced any
+-- further to fix this.
+--
+-- So, SequenceArrayRepack has to have knowledge of its actual speed
+-- in a circuit. Once again I think this complexity is worth it to
+-- contain the far greater complexity of phase matching to the
+-- SequenceArrayRepack operator.
+
+module Aetherling.Analysis.Phase (
+  boolPhase,
+  fillPhase,
+  phaseWhichCycle,
+  repackLatency
+) where
+import Data.List
+import Data.Ratio
+import Aetherling.Analysis.Metrics
+
+
+-- All of these functions are incredibly inefficient. It would be nice
+-- to try caching or a closed-form solution, but for now it is what it is.
+
+-- | Given a utilization ratio, determine the phase pattern as a list of
+-- bools (length = denominator). True indicates a valid input/output is
+-- expected on that cycle, False for garbage.
+boolPhase :: Ratio Int -> [Bool]
```

Comment:
How is it that this algorithm only does the narrowing case? Is it because the ratio must be less than 1, and so only the narrowing case is a legal input?